core/src/listener/listener.ts

```diff
@@ -220,36 +199,50 @@ function latestJob({
       observedBlock = Math.max(0, latestBlock - 1)
     }
 
-    const logPrefix = generateListenerLogPrefix(contractAddress, observedBlock, latestBlock)
     try {
       if (latestBlock > observedBlock) {
-        await redisClient.set(observedBlockRedisKey, latestBlock)
-
         // The `observedBlock` block number is already processed,
         // therefore we do not need to re-query the same event in such
         // block again.
-        const events = await state.queryEvent(contractAddress, observedBlock + 1, latestBlock)
-        for (const [index, event] of events.entries()) {
-          const outData: IProcessEventListenerJob = {
-            contractAddress,
-            event
+        const lockObservedBlock = observedBlock + 1
+        for (let blockNumber = lockObservedBlock; blockNumber <= latestBlock; ++blockNumber) {
+          const events = await state.queryEvent(contractAddress, blockNumber, blockNumber)
+          for (const [_, event] of events.entries()) {
+            const jobMetadata = await processFn(event)
+            if (jobMetadata) {
```

Comment:
how about something like this: 
```javascript
if (jobMetadata) {
  ...
} else {
  logger.debug(`jobMetadata is undefined for event ${event}`)
}
```

Reply:
I think it should be an error and not just debug log. We expect it to not be undefined or nil, therefore if it is there is something strange going on.

Reply:
Also, to be precise, we should say that the jobMetadata was not defined, rather than undefined since `undefined` is not the only case which leads to else branch.

```js
> if (!null) { console.log("hello") }
hello
undefined
> if (!undefined) { console.log("hello") }
hello
undefined
```

Reply:
Noted! then I'm throwing the following error:
```javascript
throw new Error(`jobMetadata is not defined for an event in block ${blockNumber}`)
```